File: client/pages/Resume/formatting.ts
import type React from "react";
import type { PageSize, PaperMetrics, ResumeFormatting } from "./types";
import { RESUME_RENDER_SPEC } from "./rendering/renderSpec";
import { RESUME_CSS_DEFAULTS, RESUME_CSS_DENSITY_PRESETS } from "./rendering/formattingTokens";
import { ptToPx, pxCss, pxToPt } from "./utils/documentUnits";

export const RESUME_DENSITY_PRESETS = RESUME_CSS_DENSITY_PRESETS;

export const PAPER_SIZES: Record<PageSize, PaperMetrics> = {
    a4: {
        ...RESUME_RENDER_SPEC.paperSizes.a4,
        width: ptToPx(RESUME_RENDER_SPEC.paperSizes.a4.widthPt),
        height: ptToPx(RESUME_RENDER_SPEC.paperSizes.a4.heightPt)
    },
    letter: {
        ...RESUME_RENDER_SPEC.paperSizes.letter,
        width: ptToPx(RESUME_RENDER_SPEC.paperSizes.letter.widthPt),
        height: ptToPx(RESUME_RENDER_SPEC.paperSizes.letter.heightPt)
    }
};

export const MIN_ZOOM = 0.5;
export const MAX_ZOOM = 1.4;
export const ZOOM_STEP = 0.1;
export const MIN_FIT_ZOOM = 0.25;
export const PAGE_STACK_GAP_PX = 24;


export const clampZoom = (value: number) => Math.min(MAX_ZOOM, Math.max(MIN_ZOOM, value));
export const clampFitZoom = (value: number) => Math.min(1, Math.max(MIN_FIT_ZOOM, value));

export const defaultResumeFormatting = (): ResumeFormatting => ({
    pageSize: RESUME_RENDER_SPEC.defaults.pageSize,
    paperLayoutFormat: RESUME_RENDER_SPEC.defaults.paperLayoutFormat,
    innerSectionGapFormat: RESUME_RENDER_SPEC.defaults.innerSectionGapFormat,
    ...RESUME_CSS_DEFAULTS
});


export const normalizeResumeFormatting = (formatting?: Partial<ResumeFormatting> | null): ResumeFormatting => {
    const defaults = defaultResumeFormatting();
    const pageSize = formatting?.pageSize === "letter" || formatting?.pageSize === "a4" ? formatting.pageSize : defaults.pageSize;
    const paperLayoutFormat =
        formatting?.paperLayoutFormat === "compact" || formatting?.paperLayoutFormat === "standard" || formatting?.paperLayoutFormat === "relaxed"
            ? formatting.paperLayoutFormat
            : defaults.paperLayoutFormat;
    const innerSectionGapFormat =
        formatting?.innerSectionGapFormat === "compact" || formatting?.innerSectionGapFormat === "standard" || formatting?.innerSectionGapFormat === "relaxed"
            ? formatting.innerSectionGapFormat
            : defaults.innerSectionGapFormat;

    return {
        pageSize,
        titleFontSize: clampNumberValue(formatting?.titleFontSize, ...RESUME_RENDER_SPEC.limits.titleFontSize, defaults.titleFontSize),
        headerFontSize: clampNumberValue(formatting?.headerFontSize, ...RESUME_RENDER_SPEC.limits.headerFontSize, defaults.headerFontSize),
        subHeaderFontSize: clampNumberValue(formatting?.subHeaderFontSize, ...RESUME_RENDER_SPEC.limits.subHeaderFontSize, defaults.subHeaderFontSize),
        bodyFontSize: clampNumberValue(formatting?.bodyFontSize, ...RESUME_RENDER_SPEC.limits.bodyFontSize, defaults.bodyFontSize),
        pageMarginPt: clampNumberValue(formatting?.pageMarginPt, ...RESUME_RENDER_SPEC.limits.pageMarginPt, defaults.pageMarginPt),
        paperLayoutFormat,
        innerSectionGapFormat
    };
};

export const clampNumberValue = (value: unknown, min: number, max: number, fallback: number) => {
    const numberValue = typeof value === "number" && Number.isFinite(value) ? value : fallback;
    return Math.min(max, Math.max(min, numberValue));
};

export type ResumeRenderTokens = {
    formatting: ResumeFormatting;
    paperMetrics: PaperMetrics;
    pageWidth: string;
    pageHeight: string;
    documentCssVariables: React.CSSProperties;
    pageMarginPx: number;
    pageWidthPt: number;
    pageHeightPt: number;
    pageWidthPx: number;
    pageHeightPx: number;
    contentWidthPt: number;
    contentHeightPt: number;
    contentWidth: number;
    contentHeight: number;
    titleLineHeight: number;
    headerLineHeight: number;
    subHeaderLineHeight: number;
    bodyLineHeight: number;
    titleFontSizePx: number;
    headerFontSizePx: number;
    subHeaderFontSizePx: number;
    bodyFontSizePx: number;
    sectionGapPt: number;
    innerSectionGapPt: number;
    sectionGapPx: number;
    innerSectionGapPx: number;
};

export const buildResumeRenderTokens = (
    formattingInput?: Partial<ResumeFormatting> | null,
    paperMetricsOverride?: PaperMetrics
): ResumeRenderTokens => {
    const formatting = normalizeResumeFormatting(formattingInput);
    const paperMetrics = paperMetricsOverride ?? PAPER_SIZES[formatting.pageSize];
    const paperSizeSpec = RESUME_RENDER_SPEC.paperSizes[formatting.pageSize];
    const pageWidth = paperSizeSpec.cssWidth;
    const pageHeight = paperSizeSpec.cssHeight;
    const pageWidthPt = paperMetrics.widthPt ?? pxToPt(paperMetrics.width);
    const pageHeightPt = paperMetrics.heightPt ?? pxToPt(paperMetrics.height);
    const pageWidthPx = ptToPx(pageWidthPt);
    const pageHeightPx = ptToPx(pageHeightPt);
    const pageMarginPx = ptToPx(formatting.pageMarginPt);
    const contentWidthPt = Math.max(1, pageWidthPt - formatting.pageMarginPt * 2);
    const contentHeightPt = Math.max(1, pageHeightPt - formatting.pageMarginPt * 2);
    const contentWidth = ptToPx(contentWidthPt);
    const contentHeight = ptToPx(contentHeightPt);
    const densityPreset = RESUME_DENSITY_PRESETS[formatting.paperLayoutFormat] ?? RESUME_DENSITY_PRESETS.standard;
    const titleLineHeight = densityPreset.titleLineHeight;
    const headerLineHeight = densityPreset.headerLineHeight;
    const subHeaderLineHeight = densityPreset.subHeaderLineHeight;
    const bodyLineHeight = densityPreset.bodyLineHeight;
    const sectionGapPt = densityPreset.sectionGapPt;
    const innerDensityPreset = RESUME_DENSITY_PRESETS[formatting.innerSectionGapFormat] ?? RESUME_DENSITY_PRESETS.standard;
    const innerSectionGapPt = innerDensityPreset.innerSectionGapPt;
    const sectionGapPx = ptToPx(sectionGapPt);
    const innerSectionGapPx = ptToPx(innerSectionGapPt);
    const titleFontSizePx = ptToPx(formatting.titleFontSize);
    const headerFontSizePx = ptToPx(formatting.headerFontSize);
    const subHeaderFontSizePx = ptToPx(formatting.subHeaderFontSize);
    const bodyFontSizePx = ptToPx(formatting.bodyFontSize);
    const documentCssVariables = {
        "--resume-title-font-size-pt": String(formatting.titleFontSize),
        "--resume-header-font-size-pt": String(formatting.headerFontSize),
        "--resume-subheader-font-size-pt": String(formatting.subHeaderFontSize),
        "--resume-body-font-size-pt": String(formatting.bodyFontSize),
        "--resume-page-margin-pt": String(formatting.pageMarginPt),
        "--resume-section-gap-pt": String(sectionGapPt),
        "--resume-inner-section-gap-pt": String(innerSectionGapPt),
        "--resume-title-font-size": pxCss(titleFontSizePx),
        "--resume-header-font-size": pxCss(headerFontSizePx),
        "--resume-subheader-font-size": pxCss(subHeaderFontSizePx),
        "--resume-body-font-size": pxCss(bodyFontSizePx),
        "--resume-section-gap": pxCss(sectionGapPx),
        "--resume-inner-section-gap": pxCss(innerSectionGapPx),
        "--resume-title-line-height": String(titleLineHeight),
        "--resume-header-line-height": String(headerLineHeight),
        "--resume-subheader-line-height": String(subHeaderLineHeight),
        "--resume-body-line-height": String(bodyLineHeight),
        "--resume-line-height": String(bodyLineHeight),
        "--resume-page-margin": pxCss(pageMarginPx)
    } as React.CSSProperties;

    return {
        formatting,
        paperMetrics,
        pageWidth,
        pageHeight,
        documentCssVariables,
        pageMarginPx,
        pageWidthPt,
        pageHeightPt,
        pageWidthPx,
        pageHeightPx,
        contentWidthPt,
        contentHeightPt,
        contentWidth,
        contentHeight,
        titleLineHeight,
        headerLineHeight,
        subHeaderLineHeight,
        bodyLineHeight,
        titleFontSizePx,
        headerFontSizePx,
        subHeaderFontSizePx,
        bodyFontSizePx,
        sectionGapPt,
        innerSectionGapPt,
        sectionGapPx,
        innerSectionGapPx
    };
};

File: client_api/services/resume_pdf/formatting.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from common.resume_render.spec import load_formatting_tokens, load_render_spec

RESUME_RENDER_SPEC = load_render_spec()
RESUME_FORMATTING_TOKENS = load_formatting_tokens()


@dataclass(frozen=True)
class NormalizedResumeFormatting:
    page_size: str
    page_width: str
    page_height: str
    page_name: str
    page_margin_pt: float
    title_font_size: float
    header_font_size: float
    sub_header_font_size: float
    body_font_size: float
    title_line_height: float
    header_line_height: float
    sub_header_line_height: float
    body_line_height: float
    section_gap_pt: float
    inner_section_gap_pt: float
    layout_density: str
    inner_density: str


def _read(value: Any, key: str, fallback: Any = None) -> Any:
    if isinstance(value, dict):
        return value.get(key, fallback)
    return getattr(value, key, fallback)


def _number(value: Any, minimum: float, maximum: float, fallback: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        number = fallback
    return max(minimum, min(maximum, number))


def normalize_formatting(formatting: Any) -> NormalizedResumeFormatting:
    spec = RESUME_RENDER_SPEC
    defaults = spec["defaults"]
    limits = spec["limits"]
    css_defaults = {
        "titleFontSize": RESUME_FORMATTING_TOKENS["resume-default-title-font-size-pt"],
        "headerFontSize": RESUME_FORMATTING_TOKENS["resume-default-header-font-size-pt"],
        "subHeaderFontSize": RESUME_FORMATTING_TOKENS["resume-default-subheader-font-size-pt"],
        "bodyFontSize": RESUME_FORMATTING_TOKENS["resume-default-body-font-size-pt"],
        "pageMarginPt": RESUME_FORMATTING_TOKENS["resume-default-page-margin-pt"],
    }

    page_size = _read(formatting, "pageSize", defaults["pageSize"])
    if page_size not in spec["paperSizes"]:
        page_size = defaults["pageSize"]
    paper = spec["paperSizes"][page_size]

    density_name = _read(formatting, "paperLayoutFormat", defaults["paperLayoutFormat"])
    if density_name not in spec["densityNames"]:
        density_name = defaults["paperLayoutFormat"]
    def density_token(density: str, suffix: str) -> float:
        return RESUME_FORMATTING_TOKENS[f"resume-{density}-{suffix}"]

    inner_density_name = _read(formatting, "innerSectionGapFormat", defaults["innerSectionGapFormat"])
    if inner_density_name not in spec["densityNames"]:
        inner_density_name = defaults["innerSectionGapFormat"]

    return NormalizedResumeFormatting(
        page_size=page_size,
        page_width=paper["cssWidth"],
        page_height=paper["cssHeight"],
        page_name=paper["printName"],
        page_margin_pt=_number(
            _read(formatting, "pageMarginPt", css_defaults["pageMarginPt"]),
            *limits["pageMarginPt"],
            css_defaults["pageMarginPt"],
        ),
        title_font_size=_number(
            _read(formatting, "titleFontSize", css_defaults["titleFontSize"]),
            *limits["titleFontSize"],
            css_defaults["titleFontSize"],
        ),
        header_font_size=_number(
            _read(formatting, "headerFontSize", css_defaults["headerFontSize"]),
            *limits["headerFontSize"],
            css_defaults["headerFontSize"],
        ),
        sub_header_font_size=_number(
            _read(formatting, "subHeaderFontSize", css_defaults["subHeaderFontSize"]),
            *limits["subHeaderFontSize"],
            css_defaults["subHeaderFontSize"],
        ),
        body_font_size=_number(
            _read(formatting, "bodyFontSize", css_defaults["bodyFontSize"]),
            *limits["bodyFontSize"],
            css_defaults["bodyFontSize"],
        ),
        title_line_height=density_token(density_name, "title-line-height"),
        header_line_height=density_token(density_name, "header-line-height"),
        sub_header_line_height=density_token(density_name, "subheader-line-height"),
        body_line_height=density_token(density_name, "body-line-height"),
        section_gap_pt=density_token(density_name, "section-gap-pt"),
        inner_section_gap_pt=density_token(inner_density_name, "inner-section-gap-pt"),
        layout_density=density_name,
        inner_density=inner_density_name,
    )


def format_pt(value: float) -> str:
    return f"{value:g}pt"


def format_px_from_pt(value: float) -> str:
    return f"{value * 96 / 72:g}px"


def paper_viewport_dimensions(page_name: str) -> dict[str, int]:
    spec = RESUME_RENDER_SPEC
    page_size = "letter" if page_name == "Letter" else "a4"
    paper = spec["paperSizes"][page_size]
    return {
        "width": round(paper["widthPt"] * 96 / 72),
        "height": round(paper["heightPt"] * 96 / 72),
    }
